CPack refactoring and introduction of multiversion

sfc-gh-mpilman opened the pull request:
A large refactoring of cpack:

- [x] Remove EL6 packages (resolves #3973)
- [x] Refactor ugly install destination hack
- [x] Separate utility functionality from configuration
- [x] Add multi-version package for RPM and DEB (fixes #3607)
- [x] Use alternatives to manage multiversion package

sfc-gh-mpilman commented:
I can't see why AWS CodeBuild is failing. I assume we ignore those for now?

Also: ping @sfc-gh-anoyes (as other cmake changes are in the pipeline and merging will be more complex)

xis19 commented on cmake/FDBInstall.cmake:
```diff
@@ -0,0 +1,258 @@
+endfunction()
+
+function(install_symlink_impl)
+  if (NOT WIN32)
```
nitpick: an early return could save indentation

sfc-gh-mpilman replied:
I think this is a question of taste (`return` tends to have similar problems like `goto` - I guess you can just see who my programming teacher in college was ;) ).

But I agree the indent might be a bit distracting - so I will rewrite this.

xis19 commented on cmake/FDBInstall.cmake:
```diff
@@ -0,0 +1,258 @@
+endfunction()
+
+function(install_symlink)
+  if(NOT WIN32 AND NOT OPEN_FOR_IDE)
```
Same above, I suppose

```
if (WIN32 OR OPEN_FOR_IDE)
    return()
endif()
```
could save an indentation.

xis19 commented on cmake/FDBInstall.cmake:
```diff
@@ -0,0 +1,258 @@
+        install_symlink_impl(
+          TO "../${rel_path}bin/${IN_FILE_NAME}"
+          DESTINATION "usr/lib64/${IN_LINK_NAME}"
+          COMPONENTS "${IN_COMPONENT}-el6"
```
Not sure if you want to remove the 'el6' component.

sfc-gh-mpilman replied:
see #3973 - we discussed this in our meeting two weeks ago and decided to drop el6 support as el6 is end of life for a while already.

sfc-gh-mpilman replied:
ooh, you're right Andrew, somehow I missed that...

sfc-gh-anoyes reviewed (commented):
Overall looks good. I ran into a few things attempting to install it:

1. The clients and server package seem to conflict
```
$ sudo rpm -U foundationdb-clients-6.3.10-0.1b91c99640cf8b401c7db00a6e0946e8eaeb7c49.PRERELEASE.versioned.x86_64.rpm foundationdb-server-6.3.10-0.1b91c99640cf8b401c7db00a6e0946e8eaeb7c49.PRERELEASE.versioned.x86_64.rpm
        file /usr/lib/foundationdb-6.3.10/etc/foundationdb conflicts between attempted installs of foundationdb-clients-versioned-6.3.10-1.x86_64 and foundationdb-server-6.3.10-6.3.10-1.x86_64
```

2. The name of the server rpm seems to contain the version twice: foundationdb-server-6.3.10-6.3.10-1.x86_64

3. `sudo service foundationdb start` didn't seem to work after installing
```
$ sudo service foundationdb start
Redirecting to /bin/systemctl start foundationdb.service
Failed to start foundationdb.service: Unit not found.
```

I also noticed that the server rpm doesn't depend on the client rpm, which is different from how the normal rpm works. I don't mind that so much as the use case for these rpm's isn't "get up and running for development quickly"

sfc-gh-mpilman commented:
> The clients and server package seem to conflict
I can't seem to reproduce this - but I will look into this again

> `sudo service foundationdb start` didn't seem to work after installing

This is expected behavior. Installing a service through `alternatives` doesn't seem something one can do in a portable way. For now the user would need to do this manually.

sfc-gh-mpilman commented:
@sfc-gh-anoyes I addressed all comments, but I can't reproduce the problems you ran into. Can you please try again with a clean build directory? Maybe you had an old `CPackConfig.cmake` lying around somewhere. This is how a session looks like for me:

```
[root@e7e7a0619943 foundationdb]# ls packages/*.rpm
packages/foundationdb-clients-6.3.10-0.7188c82cb02c721b84826e5265a965c45d6c02de.PRERELEASE.el7.x86_64.rpm
packages/foundationdb-clients-6.3.10-0.7188c82cb02c721b84826e5265a965c45d6c02de.PRERELEASE.versioned.x86_64.rpm
packages/foundationdb-server-6.3.10-0.7188c82cb02c721b84826e5265a965c45d6c02de.PRERELEASE.el7.x86_64.rpm
packages/foundationdb-server-6.3.10-0.7188c82cb02c721b84826e5265a965c45d6c02de.PRERELEASE.versioned.x86_64.rpm
[root@e7e7a0619943 foundationdb]# rpm -U packages/foundationdb-clients-6.3.10-0.7188c82cb02c721b84826e5265a965c45d6c02de.PRERELEASE.versioned.x86_64.rpm packages/foundationdb-server-6.3.10-0.7188c82cb02c721b84826e5265a965c45d6c02de.PRERELEASE.versioned.x86_64.rpm
[root@e7e7a0619943 foundationdb]# which fdbserver
/usr/sbin/fdbserver
```
